### ISSUE TITLE
Expand run_lm_finetuning.py to all models

### ISSUE THREAD

**Issue by iedmrc** [huggingface/transformers#2008]
## 🚀 Feature

[run_lm_finetuning.py](https://github.com/huggingface/transformers/blob/b0ee7c7df3d49a819c4d6cef977214bd91f5c075/examples/run_lm_finetuning.py) is a very useful tool for finetuning many models the library provided. But it doesn't cover all the models. Currently available models are:

- gpt2
- openai-gpt
- bert
- roberta
- distilbert
- camembert

And not available ones:

- ctrl
- xlm
- xlnet
- transfo-xl
- albert

## Motivation

Most important part of such a library is that it can be easily finetuned. `run_lm_finetuning.py` gives us that opportunity but why say no more :)

**Comment by thomwolf** [comment]
Indeed, here are my 2 cents on that:
- ctrl: easy to add (should work out of the box)
- xlm: should also work out of the box (but need to check if the model is an mlm or a clm model to finetune)
- albert: should work out of the box
- transfo-xl: need to take care of history => a little more work
- xlnet: need to take care of history + permutations => quite more work.

Do you want to give it a try? We don't have that in our short term roadmap until the end of the year.

**Comment by iedmrc** [comment]
Okay, I'm gonna try to add `ctrl`, `xlm` and `albert`. Then I'll make pull request in order to discuss on it. 

Isn't there any example of how to train `transfo-xl` and `xlnet`?

**Comment by thomwolf** [comment]
You have to look at both original repos

**Comment by CMobley7** [comment]
Out of curiosity, has any progress been made on a pull request for this?

**Comment by GabrielBianconi** [comment]
+1 for this request, especially `transfo-xl` :)

**Comment by CMobley7** [comment]
Is this issue addressed with https://github.com/huggingface/transformers/commit/a8e3336a850e856188350a93e67d77c07c85b8af?

**Comment by julien-c** [comment]
a8e3336a850e856188350a93e67d77c07c85b8af makes all those models accessible from `run_language_modeling.py`, but does not do anything special for models whose training has peculiarities, like `transfo-xl` or `xlnet`. I'm not familiar with those two so maybe someone else (@patrickvonplaten?) can chime in.

**Comment by patrickvonplaten** [comment]
As far as I know: 

Currently the `lun_language_modeling.py` script is not really made to train `transfo-xl` or `xlnet`

First as @thomwolf already said, the `mems` parameter (the "history") of the models is not taken care of during training. During training the model "caches" past sequences to effectively reuse them afterwards. It's described quite well in Figure 2 in the [Transfo-XL paper](https://arxiv.org/pdf/1901.02860.pdf). This should be rather easy to add though. 

Second, `XLNet` samples from a permutation mask during training, which is one of the core ideas of the paper, see https://github.com/huggingface/transformers/issues/2822 or Equation 5 in the official [paper](https://arxiv.org/pdf/1906.08237.pdf) This is a very special case for `XLNet` and is not yet implemented in `run_language_modeling.py` (shouldn't be too hard though to implement it since there is only one additional sum per training sample). 

Third, `Transfo-XL` uses adaptive word embeddings and adaptive softmax which also leads to some specialties when training. See also this issue #3310. This should be implemented in the model class itself though.

**Comment by CMobley7** [comment]
I'm assuming that `Albert` is fine out of the box. What about `T5`?

**Comment by urlocal12** [comment]
Is anybody still working on this currently?

**Comment by patrickvonplaten** [comment]
We are currently working on it. Might still take ~2 weeks.

**Comment by marrrcin** [comment]
Any update?

**Comment by shngt** [comment]
I'd like to try this (#4739). I'd like to start with XLNet since that's relevant to my work right now.

**Comment by patrickvonplaten** [comment]
I think you would just need to add a XLNet data collator to this file so that the trainer can be used with XLNet :-) So I would add a new XLNetLanguageModelingCollator here: https://github.com/huggingface/transformers/blob/1b5820a56540a2096daeb43a0cd8247c8c94a719/src/transformers/data/data_collator.py#L76

**Comment by shngt** [comment]
Thanks so much! I'll look into it :)

**Comment by kebegaoussou** [comment]
Any progress on XLNet? @shngt

**Comment by krannnn** [comment]
Any updates regarding XLNet ?

**Comment by shngt** [comment]
@patrickvonplaten I added the data collator as you suggested - please review :) You also mentioned earlier "the `mems` parameter (the "history") of the models is not taken care of during training" - has that been taken care of, or does the logic need to be implemented separately?

**Comment by shngt** [comment]
I was looking into the other models requested:

- CTRL -> CLM, works out of the box, already added comments
- XLM -> can be trained with three different objectives - CLM, MLM and Translation LM, which is a supervised multilingual extension of MLM. The example script does not seem to require any changes (except for maybe a warning somewhere to use the right flag with the right checkpoint?). TLM does require a lot of data-specific preprocessing, but it seems relevant only in light of the multilingual setting. I feel it would be better to incorporate those in a separate `mulitlingual_language_modeling` example script if others would like an end-to-end example of how this would be properly done.
- Albert -> Instead of the random masking in BERT, the authors use a span-based masking system first seen in SpanBERT. It is similar to what I implemented for XLNet, but the length of the span is not equally weighted (smaller spans are given preference).
- TransformerXL -> seems to be CLM with reuse of previous states. I think this functionality has been added, so no additional work should be needed

In summary, I think all that needs to be done right now for XLM and TransformerXL is to add a line or two in the starting docstring mentioning which type of LM to use. For Albert, I think we need to incorporate the masking scheme in the `DataCollator`, but am not sure about the cleanest way to do this, since the procedure also overlaps with XLNet. We could think about combining the two collators and creating a single `span_mask_tokens` procedure like `mask_tokens`. Let me know what you would prefer

@patrickvonplaten

**Comment by patrickvonplaten** [comment]
I agree very much with what you say. For `XLM` and `TransformerXL` the script should work pretty much out of the box, so we would just have to adapt some comments in `examples/language-modeling/run_language_modeling.py`.

For Albert, it would be nice to create a new `SpanMaskLanguageModeling` Data collator.

**Comment by shngt** [comment]
Great, I'll get started then. I'll try to finish it over the weekend :)

**Comment by patrickvonplaten** [comment]
Awesome, no rush though ;-)

**Comment by KristenMoore** [comment]
Maybe a stupid question, but where should I find `run_lm_finetuning.py`? The docs point to a [dead link,](https://huggingface.co/transformers/v2.0.0/examples.html) as the file doesn't exist in the master branch.

**Comment by iedmrc** [comment]
it's renamed and moved [there](https://github.com/huggingface/transformers/tree/master/examples/language-modeling).

**Comment by patrickvonplaten** [comment]
Thanks for the notice @KristenMoore - I will update the docs.
